### DISCUSSION TITLE
How to override group and group variable at the same time

### DISCUSSION THREAD

**Discussion by camall3n** [facebookresearch/hydra#2391] [Q&A] [answered]
# 🐛 Bug
## Description
Cannot override group and group variable at the same time.

For example:
```text
 python main.py car=vw car/color=black
```

## To reproduce

Directory structure:
```text
├── conf
│   ├── car
│   │   └── vw.yaml
│   └── config.yaml
└── main.py
```

main.py:
```python
import hydra

@hydra.main(config_path="./conf", config_name='config', version_base=None)

def main(cfg):
    pass

main()
```

conf/config.yaml:
```yaml
defaults:
  - _self_
  - car: null

name: Alice
car: ???
```

conf/car/vw.yaml:
```yaml
make: vw
model: beetle
color: red
```

I can print the default config and optionally override the `car` with `vw` or a dictionary that sets the `color` field, but when I try to set both the `car=vw` and the `car/color=black` at the same time, it fails.
```text
$ python main.py --cfg job
name: Alice
car: ???

$ python main.py car=vw --cfg job
name: Alice
car:
  make: vw
  model: beetle
  color: red

$ python main.py car="{make:bmw,model:3series,color:black}" --cfg job
name: Alice
car:
  make: bmw
  model: 3series
  color: black

$ python main.py car=vw car/color=black --cfg job
Could not override 'car/color'.
To append to your config use +car/color=black
Key 'car/color' is not in struct
    full_key: car/color
    object_type=dict

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
```

** Stack trace/error message **
```
Traceback (most recent call last):
...
    raise ex.with_traceback(sys.exc_info()[2])  # set env var OC_CAUSE=1 for full trace
omegaconf.errors.ConfigAttributeError: Key 'car/color' is not in struct
    full_key: car/color
    object_type=dict

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
...
hydra.errors.ConfigCompositionException: Could not override 'car/color'.
To append to your config use +car/color=black
```

## Expected Behavior
The `car` group info should come from vw.yaml, and the `color` field should be 'black':
```text
$ python main.py car=vw --cfg job
name: Alice
car:
  make: vw
  model: beetle
  color: black
```

**Note:** Overriding the variable using `+car/color=black` also doesn't work:
```text
$ python main.py car=vw +car/color=black --cfg job
name: Alice
car:
  make: vw
  model: beetle
  color: red
car/color: black
```

## System information
- **Hydra Version** : 1.3.0.dev - 92a53421
- **Python version** : 3.9.13
- **Virtual environment type and version** : python venv + pip (22.1.1)
- **Operating system** : macOS 10.15.7

**Comment by Jasha10** [comment, marked as answer]
Give this a try:
```python
python main.py car=vw car.color=black
```

The syntax `'foo/bar=value'` is used for modifying the defaults list (e.g. for appending to the defaults list or overriding an existing entry in the defaults list). Meanwhile, the syntax `'foo.bar=value'` directly modifies the composed config object.

Ref: [Hydra's Override grammar](https://hydra.cc/docs/1.2/advanced/override_grammar/basic/#basic-examples)

**Comment by camall3n** [comment]
Ah, that works! Thank you.

**Comment by charleswilmot** [comment]
I do observe the problem as described by @camall3n , when one of the defaults is set to `???` and I use the `multirun` sweep feature.
I did pay attention to using dots and not slashes like @camall3n did at first.


**config.yaml**
```
defaults:
    - something=???
```

**something/option1.yaml**
```
param1: 123
param2: abc
```

**something/option2.yaml**
```
param1: 123
param2: abc
```

command line:
```
python my_app.py -m something=option1,option2 param1=456 param2=def
```
results in
```
Could not override 'something.param1'.
To append to your config use +something.param1=456
Key 'experiment' is not in struct
```

command line:
```
python my_app.py -m something=option1,option2 +param1=456 +param2=def
```
results in
```
Could not append to config. An item is already at 'something.param2'.
Either remove + prefix: 'something.param2=def'
Or add a second + to add or override 'something.param2': '++something.param2=def'
```

**The only solution that works in this case is to run the command**
```
python my_app.py -m something=option1,option2 +param1=456 ++param2=def
```

where the first parameter is prepended with 1 `+` and all the next parameters with two `++`

@Jasha10

**Reply by Jasha10** [reply]
I believe what you're seeing is due to a subtle limitation in the way that Hydra executes multirun sweeps; hydra composes the multirun config in two phases, first composing a config that is common between all the multirun jobs and then doing an additional round of composition (specializing the config) for each job in the sweep. See [my comment here](https://github.com/facebookresearch/hydra/discussions/2329#discussioncomment-3612988) for more details on this.

**Reply by charleswilmot** [reply]
Thank you for the answer
